File: src/skimr/extract/metadata.py
"""extract.metadata — stdlib core + pluggable backend dispatch.

Core path (always available):
  - dates: ISO yyyy-mm-dd and US m/d/yyyy
  - amounts: money patterns (similar shape to extract.stats, defined locally)
  - urls: http(s) URLs

Backend selection:
  - backend='regex' (default): zero-dep stdlib regex path. entities always ().
  - backend='spacy': requires the skimr-spacy companion package; it registers
    itself on import and populates entities via en_core_web_sm.
  - backend='auto': spacy if registered, else regex.
  - backend=None: use the process-wide default (skimr.set_default_backend).

See docs/superpowers/specs/2026-04-21-skimr-spacy-integration.md.
"""
import re
from .._types import Metadata
from . import _backends


# Date patterns mirror extract.stats._DATE_RE: ISO yyyy-mm-dd, US m/d/yyyy,
# and bare years 1900-2099. Same shape, same bounds, same parity contract.
_DATE_RE = re.compile(
    r"\b\d{4}-\d{2}-\d{2}\b|\b\d{1,2}/\d{1,2}/\d{2,4}\b|\b(?:19|20)\d{2}\b"
)
# Amount quantifiers bounded {0,18} for ReDoS parity with extract.stats._MONEY_RE.
_AMOUNT_RE = re.compile(
    r"\$\d[\d,]{0,18}(?:\.\d{1,4})?[KMB]?|\d[\d,]{0,18}(?:\.\d{1,4})?\s*(?:dollars?|USD|EUR|GBP|JPY|CHF)",
    re.IGNORECASE,
)
_URL_RE = re.compile(
    r"https?://[^\s<>\"')]+",
    re.IGNORECASE,
)


def _collect_unique(regex: re.Pattern, text: str) -> tuple[str, ...]:
    """Find all matches preserving first-appearance order, deduped."""
    seen: set[str] = set()
    out: list[str] = []
    for m in regex.finditer(text):
        v = m.group(0)
        if v not in seen:
            seen.add(v)
            out.append(v)
    return tuple(out)


def _regex_metadata(text: str) -> Metadata:
    """Regex backend: zero-dep core path. entities always ()."""
    if not text:
        return Metadata()
    return Metadata(
        dates=_collect_unique(_DATE_RE, text),
        amounts=_collect_unique(_AMOUNT_RE, text),
        urls=_collect_unique(_URL_RE, text),
        entities=(),
    )


# Regex backend self-registers at module load so it is always available.
_backends.register("regex", "metadata", _regex_metadata)


def metadata(text: str, *, backend: str | None = None) -> Metadata:
    """Core metadata + optional neural entities via backend= kwarg.

    backend='regex' (default): zero-dep regex path. Populates dates/amounts/urls;
        entities always empty.
    backend='spacy': requires skimr-spacy installed. Populates entities too.
    backend='auto': spacy if registered else regex.
    backend=None: uses the global default (skimr.set_default_backend).
    """
    if backend is None:
        backend = _backends.get_default_backend()
    impl = _backends.resolve(backend, "metadata")
    return impl(text)

File: rust/src/extract/metadata.rs
//! Metadata extractor — stdlib core only; entities always empty in Rust.

use regex::Regex;
use std::sync::OnceLock;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct Metadata {
    pub dates: Vec<String>,
    pub amounts: Vec<String>,
    pub urls: Vec<String>,
    pub entities: Vec<String>, // always empty; Python-only via lede[ner]
}

// Date patterns mirror extract::stats: ISO yyyy-mm-dd, US m/d/yyyy, and bare
// years 1900-2099. Same shape, same bounds — keeps Python ↔ Rust output
// byte-identical and matches the doc claim in docs/REFERENCE.md.
fn date_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"\b\d{4}-\d{2}-\d{2}\b|\b\d{1,2}/\d{1,2}/\d{2,4}\b|\b(?:19|20)\d{2}\b")
            .expect("static regex")
    })
}

// Amount quantifiers bounded {0,18} for parity with extract::stats::money_re.
fn amount_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(concat!(
            r"(?i)\$\d[\d,]{0,18}(?:\.\d{1,4})?[KMB]?(?:\s+(?:million|billion|trillion|thousand)\b)?|",
            r"\d[\d,]{0,18}(?:\.\d{1,4})?(?:\s+(?:million|billion|trillion|thousand)\b)?\s*(?:dollars?|USD|EUR|GBP|JPY|CHF)",
        ))
        .expect("static regex")
    })
}

fn url_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"(?i)https?://[^\s<>"')]+"#).expect("static regex"))
}

fn collect_unique(re: &Regex, text: &str) -> Vec<String> {
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut out: Vec<String> = Vec::new();
    for m in re.find_iter(text) {
        if !seen.contains(m.as_str()) {
            let v = m.as_str().to_string();
            seen.insert(v.clone());
            out.push(v);
        }
    }
    out
}

#[must_use]
pub fn metadata(text: &str) -> Metadata {
    if text.is_empty() {
        return Metadata::default();
    }
    Metadata {
        dates: collect_unique(date_re(), text),
        amounts: collect_unique(amount_re(), text),
        urls: collect_unique(url_re(), text),
        entities: Vec::new(),
    }
}
